### ISSUE TITLE
Improve load times by adding build option to use CDN version of Bootstrap

### ISSUE THREAD

**Issue by henrify** [roots/sage#1549]
Lot of end users already have bootstrap cached from CDN.

I believe that there would be significant performance improvement in load time of the first page if you could provide alternative not to bundle/combine Bootstrap to main.css, but instead link to CDN version.

Of course this also means that Bootstrap's Sass custom builds are not available, but for many this is perfectly fine (e.g. i prefer to use 100% vanilla bootstrap anyway).

If you decline adding the proposed build option, perhaps you could at least document how to achieve the same thing manually by modifying build setup.

**Comment by JulienMelissas** [comment]
> Lot of end users already have bootstrap cached from CDN.

That's an assumption I wouldn't be willing to make personally. Do you have any data for this?

> Of course this also means that Bootstrap's Sass custom builds are not available, but for many this is perfectly fine

Again, an assumption I wouldn't be willing to make, especially because almost every single job I personally do with Sage involves changing around BS variables (such as `$brand-primary` or `$grid-gutter-width`).

I'm going to :-1: this one, but I'm happy to document how one would remove Bootstrap as a dependency and instead choose the CDN:

### Remove Bootstrap from Bower:
Delete [this line](https://github.com/roots/sage/blob/master/bower.json#L11), and delete [these lines](https://github.com/roots/sage/blob/master/bower.json#L16-L37).

Run `gulp` and wiredep will remove the Boostrap references from main.scss (https://github.com/roots/sage/blob/master/assets/styles/main.scss#L5)

### Remove Bootstrap references from the Scss files:
Remove everything in [`_variables`](https://github.com/roots/sage/blob/master/assets/styles/common/_variables.scss).
Remove references in ['_comments'](https://github.com/roots/sage/blob/master/assets/styles/components/_comments.scss) - but remember your comments are going to look pretty blank.
Remove references in ['_wp-classes'](https://github.com/roots/sage/blob/master/assets/styles/components/_wp-classes.scss) - even more blank looking now...
Remove references in ['_grid'](https://github.com/roots/sage/blob/master/assets/styles/components/_grid.scss) - but you'll have to manually add classes to the sidebar and the main div [here](https://github.com/roots/sage/blob/master/base.php#L23) and [here](https://github.com/roots/sage/blob/master/base.php#L27).

### Enqueue Bootstrap from a CDN:
Add these lines directly above where we (enqueue the Sage CSS and JS)[https://github.com/roots/sage/blob/master/lib/assets.php#L69]:
```
wp_enqueue_style('bootstrap/css', '//maxcdn.bootstrapcdn.com/bootstrap/3.3.5/js/bootstrap.min.css', false, null);
wp_enqueue_script('bootstrap/js', '//maxcdn.bootstrapcdn.com/bootstrap/3.3.5/js/bootstrap.min.js', ['jquery'], null, true);
```

## If you take a look, I don't think it's worth it...
You're going to end up making a whole lot more work for yourself IMO, and on top of that, we've just added 2 extra requests. Is that worth the performance boost from the chance that they might be loading from a CDN and the fact that now you have to manually add BS classes to lots of WP elements?

**Comment by chrisatomix** [comment]
> I believe that there would be significant performance improvement in load time of the first page if you could provide alternative not to bundle/combine Bootstrap to main.css, but instead link to CDN version.

Every performance test I've ever run tells me this is not correct, because of the additional DNS lookups and latency required. In many cases you'll actually see pages hanging because "waiting for ajax.googleapis.com" (or whatever the CDN is).

The CSS file on the site I'm currently working on is 208KB, that includes Bootstrap (the entire SCSS file with some overridden variables), a lightbox plugin and all our custom CSS.
The compiled file is 208KB, but a properly configured server reduces this down to only 45KB with gzip compression.
Our performance grade on the site has improved drastically too, from a "D" to an "A" grade (in ySlow). Replacing "Share This" with the Roots Share Icons plugin will boost our score further, as it adds a bunch of unnecessary JS files to the page.

Try profiling your site's performance both before and after making the above changes and see which one performs better for you (disable your browser cache if necessary, or use something like ySlow or Pingdom). I'd be surprised if CDN Bootstrap loaded faster than a compiled/compressed version, but the great thing about Sage is that it's flexible enough to be able to override anything for your own needs :)

**Comment by henrify** [comment]
Thanks for your fast replies. I got bootstrap chopped off with the instructions, though I haven't fixed the styling yet.

Btw to avoid making non-fact-backed assumptions, I specifically said 'many' users, which considering the scale of internet is likely true :)

**Comment by henrify** [comment]
> Every performance test I've ever run tells me this is not correct, because of the additional DNS lookups and latency required

You are talking about situations where the resources are not cached, right? For that I don't doubt what you are saying.

But if the resources are cached in the end user's browser and the cache headers have been set properly there won't be any dns lookup or http request at all, and the resulting 0ms load time is hard to beat.

**Comment by kalenjohnson** [comment]
> But if the resources are cached in the end user's browser and the cache headers have been set properly there won't be any dns lookup or http request at all, and the resulting 0ms load time is hard to beat.

And either way, after the initial page load, all CSS is cached. You're giving up a whole lot of power and customizability for a gamble at saving milliseconds of load time.

**Comment by chrisatomix** [comment]
What percentage of visitors would have your exact version of vanilla Bootstrap cached? I doubt it could be anywhere higher than 15-20%, and even that seems like an extremely high number.

If the percentage was correct, that would mean that the other 80% of your visitors won't have Bootstrap cached, and will need to perform that additional lookup & download.

In the end it's your call, you know your visitors/audience better than we do. For me personally I'd prefer to reduce the number of assets for the 80% of my visitors. One CSS file in the header and one JS file in the footer, it's lightning fast! :)

**Comment by QWp6t** [comment]
No perceivable performance boost, even if it is cached.

Still gotta load local CSS regardless of whether Bootstrap is cached. Local CSS might be 5K instead of 50K if Bootstrap is removed, but the difference between the two is imperceivable.

This is micro-optimizing at best. To call it a "significant performance improvement" is absurd.

**Comment by henrify** [comment]
> Every performance test I've ever run tells me this is not correct, because of the additional DNS  lookups and latency required

This didn't turn out to be so clear cut after all. For clients that are further away from host it seems that CDN starts to gain advantage as they can serve the files from a nearby PoP.

Maybe keeping bootstrap and jquery combined as in default setup but placing main.js/css to CDN would be reasonable compromise.

**Comment by henrify** [comment]
> No perceivable performance boost, even if it is cached.

I tested this from several locations globally, and the measured effect of jquery + bootstrap being cached was up to 2.5 seconds.

If you do your testing sloppily e.g. just open web browser in a client close to the host you probably won't see "perceived" difference.

> This is micro-optimizing at best. To call it a "significant performance improvement" is absurd.

The effect can be very significant in some situations as outlined above. No need to be arrogant.

**Comment by henrify** [comment]
> What percentage of visitors would have your exact version of vanilla Bootstrap cached? I doubt it could be anywhere higher than 15-20%, and even that seems like an extremely high number.

That question is good, important and unfortunately hard to answer. After quick thinking, i'm actually not even sure how to collect measurements whether a resource came from cache or CDN, any idea?

**Comment by retlehs** [comment]
feel free to move this discussion to our discourse but i'm locking this convo.
